### PR TITLE
[VP][RISCV] Add llvm.experimental.vp.reverse.

### PR DESCRIPTION
**Pull request by topperc**
This is similar to vector.reverse, but only reverses the first EVL elements.

I extracted this code from our downstream. Some of it may have come from https://repo.hca.bsc.es/gitlab/rferrer/llvm-epi/ originally.

### CODE REVIEW

File: llvm/include/llvm/IR/VPIntrinsics.def [hunk 1]
```diff
@@ -702,6 +702,12 @@ BEGIN_REGISTER_VP(experimental_vp_splice, 3, 5, EXPERIMENTAL_VP_SPLICE, -1)
 VP_PROPERTY_FUNCTIONAL_INTRINSIC(experimental_vector_splice)
 END_REGISTER_VP(experimental_vp_splice, EXPERIMENTAL_VP_SPLICE)
 
+// llvm.experimental.vp.reverse(x,mask,vlen)
+BEGIN_REGISTER_VP(experimental_vp_reverse, 1, 2,
+                  EXPERIMENTAL_VP_REVERSE, -1)
+VP_PROPERTY_FUNCTIONAL_INTRINSIC(experimental_vector_reverse)
```
Comment (lukel97) [inline llvm/include/llvm/IR/VPIntrinsics.def:708]: The expand vector predication pass will expand out llvm.experimental.vp.reverse to llvm.experimental.vector.reverse and drop the EVL if we mark it as functionally equivalent, but without the EVL it won't be the same reverse. Should this be marked as `VP_PROPERTY_NO_FUNCTIONAL` instead?

Comment (topperc) [reply llvm/include/llvm/IR/VPIntrinsics.def:708]: Does a similar issue exist for vp.splice, or does expand vector predication do something special?

File: llvm/lib/Target/RISCV/RISCVISelLowering.cpp [hunk 2]
```diff
@@ -10223,6 +10231,143 @@ SDValue RISCVTargetLowering::lowerVPFPIntConvOp(SDValue Op,
...
+    case 32:
+    case 64: IndicesVT = MVT::i8; break;
+    }
+    GatherVT = IndicesVT = ContainerVT.changeVectorElementType(IndicesVT);
+
+    // Expand input operand
+    SDValue SplatOne = DAG.getNode(RISCVISD::VMV_V_X_VL, DL, IndicesVT,
+                                   DAG.getUNDEF(IndicesVT),
+                                   DAG.getConstant(1, DL, XLenVT), EVL);
+    SDValue VMV0 = DAG.getNode(RISCVISD::VMV_V_X_VL, DL, IndicesVT,
```
Comment (frasercrmck) [inline llvm/lib/Target/RISCV/RISCVISelLowering.cpp:10274]: `SplatOne` and `VMV0` could maybe more similarly named.

File: llvm/lib/Target/RISCV/RISCVISelLowering.cpp [hunk 3]
```diff
@@ -10223,6 +10231,143 @@ SDValue RISCVTargetLowering::lowerVPFPIntConvOp(SDValue Op,
...
+  // NOTE: This code assumes VLMAX <= 65536 for LMUL=8 SEW=16.
+  if ((MaxVLMAX == 0 || MaxVLMAX > 256) && EltSize == 8) {
+    // If this is LMUL=8, we have to split before using vrgatherei16.vv.
+    // Split the vector in half and reverse each half using a full register
+    // reverse.
+    // Swap the halves and concatenate them.
+    // Slide the concatenated result by (VLMax - VL).
+    if (MinSize == (8 * RISCV::RVVBitsPerBlock)) {
+      EVT LoVT, HiVT;
+      std::tie(LoVT, HiVT) = DAG.GetSplitDestVTs(GatherVT);
```
Comment (frasercrmck) [inline llvm/lib/Target/RISCV/RISCVISelLowering.cpp:10303]: I'm a bit out of the loop - are we still using this form, or is `auto [LoVT, HiVT] =` okay?

File: llvm/include/llvm/IR/VPIntrinsics.def (hunk 1 above)
Comment (lukel97) [reply llvm/include/llvm/IR/VPIntrinsics.def]: Yeah good catch, I'll post a PR to fix that

File: llvm/lib/Target/RISCV/RISCVISelLowering.cpp [hunk 4]
```diff
@@ -10223,6 +10231,141 @@ SDValue RISCVTargetLowering::lowerVPFPIntConvOp(SDValue Op,
...
+                                    DAG.getConstant(0, DL, XLenVT), EVL);
+    Op1 = DAG.getNode(RISCVISD::VSELECT_VL, DL, IndicesVT, Op1, SplatOne,
+                      SplatZero, EVL);
+  }
+
+  unsigned EltSize = GatherVT.getScalarSizeInBits();
+  unsigned MinSize = GatherVT.getSizeInBits().getKnownMinValue();
+  unsigned MaxVLMAX = 0;
+  unsigned VectorBitsMax = Subtarget.getRealMaxVLen();
+  if (VectorBitsMax != 0)
```
Comment (lukel97) [inline llvm/lib/Target/RISCV/RISCVISelLowering.cpp:10285]: When do we get a max VLEN of 0?

Comment (topperc) [reply llvm/lib/Target/RISCV/RISCVISelLowering.cpp:10285]: We don't anymore. This code has been in our downstream for a long time. I'll remove it to match lowerVECTOR_REVERSE.

File: llvm/lib/Target/RISCV/RISCVISelLowering.cpp [hunk 5]
```diff
@@ -10223,6 +10231,129 @@ SDValue RISCVTargetLowering::lowerVPFPIntConvOp(SDValue Op,
...
+  unsigned MaxVLMAX =
+      RISCVTargetLowering::computeVLMAX(VectorBitsMax, EltSize, MinSize);
+
+  unsigned GatherOpc = RISCVISD::VRGATHER_VV_VL;
+  // If this is SEW=8 and VLMAX is unknown or more than 256, we need
+  // to use vrgatherei16.vv.
+  // TODO: It's also possible to use vrgatherei16.vv for other types to
+  // decrease register width for the index calculation.
+  // NOTE: This code assumes VLMAX <= 65536 for LMUL=8 SEW=16.
+  if (MaxVLMAX > 256 && EltSize == 8) {
```
Comment (lukel97) [inline llvm/lib/Target/RISCV/RISCVISelLowering.cpp:10283]: Just an idea, not for this patch, but could we avoid this for fixed length vectors if we know the length is < 256, or if EVL is a constant and we know it's < 256?

File: llvm/lib/Target/RISCV/RISCVISelLowering.cpp [hunk 6]
```diff
@@ -10223,6 +10231,129 @@ SDValue RISCVTargetLowering::lowerVPFPIntConvOp(SDValue Op,
...
+      // elements.
+      unsigned MinElts = GatherVT.getVectorMinNumElements();
+      SDValue VLMax = DAG.getNode(ISD::VSCALE, DL, XLenVT,
+                                  DAG.getConstant(MinElts, DL, XLenVT));
+      SDValue Diff = DAG.getNode(ISD::SUB, DL, XLenVT, VLMax, EVL);
+
+      SDValue TrueMask = getAllOnesMask(ContainerVT, EVL, DL, DAG);
+      Result =
+          getVSlidedown(DAG, Subtarget, DL, GatherVT, DAG.getUNDEF(GatherVT),
+                        Result, Diff, TrueMask, EVL);
```
Comment (lukel97) [inline llvm/lib/Target/RISCV/RISCVISelLowering.cpp:10313]: Does using the input mask here like
```suggestion
                        Result, Diff, Mask, EVL);
```
work? So we get "poison" in the expected lanes

File: llvm/lib/Target/RISCV/RISCVISelLowering.cpp (hunk 5 above)
Comment (topperc) [reply llvm/lib/Target/RISCV/RISCVISelLowering.cpp:10283]: I think so. Do you know of anyone using VP intrinsics on fixed vectors today? Or using VP intrinsics with constant VL?

Comment (lukel97) [reply llvm/lib/Target/RISCV/RISCVISelLowering.cpp:10283]: Off the top of my head nothing comes to mind

**Review by lukel97 — APPROVED** [review]
LGTM

**Pull request merged**